### ISSUE TITLE
trying to use resnet50 as a backbone head yolov8

### ISSUE THREAD

**Issue by SupriyaBorhade** [ultralytics/ultralytics#21493]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

 #Parameters
nc: 80  # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]
  s: [0.33, 0.50, 1024]
  m: [0.67, 0.75, 768]
  l: [1.00, 1.00, 512]
  x: [1.00, 1.25, 512]

backbone:
  # [from, repeats, module, args]
  # ResNetLayer(args: input_channels, output_channels, stride, is_first:bool, n_blocks)
  - [-1, 1, ResNetLayer, [3, 64, 1, True, 1]] # 0 P1/2 |  output_size: 64, 160, 160
  - [-1, 1, ResNetLayer, [64, 64, 1, False, 3]] # 1 P2/4 | output_size: 256, 160, 160
  - [-1, 1, ResNetLayer, [256, 128, 2, False, 4]] # 2 P3/8 | output_size: 512, 80, 80
  - [-1, 1, ResNetLayer, [512, 256, 2, False, 6]] # 3 P4/16 | output_size: 1024, 40, 40
  - [-1, 1, ResNetLayer, [1024, 512, 2, False, 3]] # 4 P5/32 | output_size: 2048, 20, 20
  - [-1, 1, SPPF, [1024, 5]]  # 5 | output_size: 512, 20, 20
  
# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]  # 6  output_size: 512, 40, 40
  - [-1, 1, Conv, [1024, 3, 2]] # output_size: 1024, 40, 40
  - [[-1, 3], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 9

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 2], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 12 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 7], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 15 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 18 (P5/32-large)

  - [[12, 15, 18], 1, Detect, [nc]]  # Detect(P3, P4, P5)          
(D:\conda_envs\yolov8_custom) D:\model2>python -m yolo_res.tests.test_full_model1

✅ Testing Full Model Integration...

🔍 Backbone Output Shapes:
  Feature 0: torch.Size([1, 256, 160, 160])
  Feature 1: torch.Size([1, 512, 80, 80])
  Feature 2: torch.Size([1, 1024, 40, 40])

                   from  n    params  module                                       arguments
  0                  -1  1     39168  yolo_res.models.resnet_blocks.ResNetLayer    [3, 64, 1, True, 1]
  1                  -1  1    221952  yolo_res.models.resnet_blocks.ResNetLayer    [64, 64, 1, False, 3]
  2                  -1  1   1362176  yolo_res.models.resnet_blocks.ResNetLayer    [256, 128, 2, False, 4]
  3                  -1  1   7805440  yolo_res.models.resnet_blocks.ResNetLayer    [512, 256, 2, False, 6]
  4                  -1  1  17046528  yolo_res.models.resnet_blocks.ResNetLayer    [1024, 512, 2, False, 3]
  5                  -1  1   3148288  ultralytics.nn.modules.block.SPPF            [2048, 256, 5]
  6                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
  7                  -1  1    590336  ultralytics.nn.modules.conv.Conv             [256, 256, 3, 2]
  8             [-1, 3]  1         0  ultralytics.nn.modules.conv.Concat           [1]
  9                  -1  1    262912  ultralytics.nn.modules.block.C2f             [1280, 128, 1]
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 11             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 12                  -1  1     65920  ultralytics.nn.modules.block.C2f             [640, 64, 1]
 13                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]
 14             [-1, 7]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 15                  -1  1    140032  ultralytics.nn.modules.block.C2f             [320, 128, 1]
 16                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]
 17             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 18                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1]
 19        [12, 15, 18]  1    897664  ultralytics.nn.modules.head.Detect           [80, [64, 128, 256]]

❌ Model loading or forward pass failed: Given groups=1, weight of size [128, 256, 3, 3], expected input[1, 64, 256, 256] to have 256 channels, but got 64 channels instead

<img width="1882" height="1011" alt="Image" src="https://github.com/user-attachments/assets/5b8ef78f-2262-46b6-93f6-8184a1b67f0e" />

### Environment

Python version: 3.9.23
Ultralytics 8.3.169  Python-3.9.23 torch-2.5.1 CUDA:0 (NVIDIA GeForce RTX 4050 Laptop GPU, 6140MiB)
Setup complete  (12 CPUs, 15.6 GB RAM, 11.4/223.1 GB disk)

OS                  Windows-10-10.0.26100-SP0
Environment         Windows
Python              3.9.23
Install             pip
Path                D:\conda_envs\yolov8_custom\Lib\site-packages\ultralytics
RAM                 15.65 GB
Disk                11.4/223.1 GB
CPU                 13th Gen Intel Core(TM) i5-13420H
CPU count           12
GPU                 NVIDIA GeForce RTX 4050 Laptop GPU, 6140MiB
GPU count           1
CUDA                11.8

numpy                2.0.1>=1.23.0
matplotlib           3.9.4>=3.3.0
opencv-python        4.12.0.88>=4.6.0
pillow               11.1.0>=7.1.2
pyyaml               6.0.2>=5.3.1
requests             2.32.4>=2.23.0
scipy                1.13.1>=1.4.1
torch                2.5.1>=1.8.0
torch                2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision          0.20.1>=0.9.0
tqdm                 4.67.1>=4.64.0
psutil               7.0.0
py-cpuinfo           9.0.0
pandas               2.3.1>=1.1.4
ultralytics-thop     2.0.14>=2.0.0

### Minimal Reproducible Example

<img width="1882" height="1011" alt="Image" src="https://github.com/user-attachments/assets/0717be21-cdd5-4b7e-a650-ac4d66d9774e" />

### Additional

please help in this 

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SupriyaBorhade, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. It looks like you’ve already shared some details and screenshots—thank you! If possible, please include a concise script or notebook that others can use to reproduce your exact error.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will review your issue and assist you soon!

**Comment by Gabrli** [comment]
Hi,

I noticed in layer number 16 - you have 128 channels, but it should be 256, so problem is somewhere in yaml configuration.

**Comment by glenn-jocher** [comment]
@Gabrli is correct - there's a channel mismatch in your YAML configuration. The issue is that ResNet50 uses bottleneck blocks that expand channels by 4x, but your configuration doesn't account for this.

Here's the corrected backbone configuration for ResNet50:

```yaml
backbone:
  - [-1, 1, ResNetLayer, [3, 64, 1, True, 1]]      # 64 channels
  - [-1, 1, ResNetLayer, [64, 64, 1, False, 3]]    # 256 channels (4x expansion)
  - [-1, 1, ResNetLayer, [256, 128, 2, False, 4]]  # 512 channels (4x expansion)
  - [-1, 1, ResNetLayer, [512, 256, 2, False, 6]]  # 1024 channels (4x expansion) 
  - [-1, 1, ResNetLayer, [1024, 512, 2, False, 3]] # 2048 channels (4x expansion)
  - [-1, 1, SPPF, [2048, 1024, 5]]  # Input should be 2048, output 1024
```

Also update your head to use the correct channel counts from the backbone outputs. The detection head channels should match: `[256, 512, 1024]` for P3, P4, P5 respectively.

**Comment by binhere** [comment]
Hi @glenn-jocher, I have some questions about custom architecture via yaml file:

1. When using the ResNetLayer as a backbone, does it use ImageNet-pretrained weights for the backbone and still keep the YOLOv8.0n head's weights?
2. The ResNetLayer is a backbone of Resnet50, right? Are there any additional options, such as ResNet18/Resnet101 ?

Thanks for your support.

**Comment by glenn-jocher** [comment]
Great questions:

1) Using a custom YAML with `ResNetLayer` won’t automatically load ImageNet weights; only parameters that match a supplied checkpoint are loaded, so swapping in a ResNet backbone typically invalidates the YOLOv8n head shapes and those layers are reinitialized. If you need pretrained features, provide a checkpoint for this exact architecture or consider integrating a torchvision backbone via the `TorchVision` wrapper noted in the YAML parser docs; see the parse logic in the [parse_model reference](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.parse_model).

2) `ResNetLayer` implements bottleneck-style stages (`e=4`), suitable for building ResNet‑50/101 layouts by selecting appropriate repeats per stage; basic-block variants like ResNet‑18/34 aren’t provided out of the box. Details are in the [ResNetLayer reference](https://docs.ultralytics.com/reference/nn/modules/block/#ultralytics.nn.modules.block.ResNetLayer).

**Comment by binhere** [comment]
Thank you @glenn-jocher for your clarification, but as you said:

> 1. Using a custom YAML with `ResNetLayer` won’t automatically load ImageNet weights; only parameters that match a supplied checkpoint are loaded, so swapping in a ResNet backbone typically invalidates the YOLOv8n head shapes and those layers are reinitialized. 

In other words, **among all models provided by Ultralytics**, backbone customization via YAML only reconstructs the backbone's architecture that matches the remaining parts of the model, **not leveraging and reusing** pre-trained weights, right?

For example, there are many architectures that I found in [Ultralytics Repo](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/cfg/models)

- yolov8-cls-resnet50.yaml -> just build model's backbone like the resnet50 architecture, the weights of new backbone are reinitialized (not pre-trained weights on ImageNet)
- yolov8-ghost-p2.yaml -> reinitialize weights of ghost backbone
- rtdetr-resnet50.yaml -> reinitialize weights of resnet50 backbone

**Comment by glenn-jocher** [comment]
@binhere Correct—YAML only defines architecture; weights are new unless you load a matching checkpoint and transfer overlapping layers. For pretrained starts, use a .pt file that matches your model, for example:

```python
from ultralytics import YOLO
model = YOLO("custom.yaml").load("some_pretrained.pt")  # loads matching keys; others reinit
```

or CLI:

```bash
yolo detect train model=custom.yaml pretrained=some_pretrained.pt
```

See the Train mode examples on building from YAML and transferring weights in the [training docs](https://docs.ultralytics.com/modes/train/), and when official weights exist (e.g., RT-DETR), prefer the corresponding .pt models shown on the [RT-DETR page](https://docs.ultralytics.com/models/rtdetr/).
